src/libaktualizr/package_manager/dockerappmanager.cc

```diff
@@ -0,0 +1,99 @@
+#include "dockerappmanager.h"
+
+#include <sstream>
+
+struct DockerApp {
+  DockerApp(const std::string app_name, const PackageConfig &config)
+      : name(std::move(app_name)),
+        app_root(std::move(config.docker_apps_root / app_name)),
+        app_params(std::move(config.docker_app_params)),
+        app_bin(std::move(config.docker_app_bin)),
+        compose_bin(std::move(config.docker_compose_bin)) {}
+
+  bool render(const std::string &app_content) {
+    auto bin = boost::filesystem::canonical(app_bin).string();
+    Utils::writeFile(app_root / (name + ".dockerapp"), app_content);
+    std::string cmd("cd " + app_root.string() + " && " + bin + " app render " + name);
+    if (!app_params.empty()) {
+      cmd += " -f " + app_params.string();
+    }
+    std::string yaml;
+    if (Utils::shell(cmd, &yaml, true) != 0) {
+      LOG_ERROR << "Unable to run " << cmd << " output:\n" << yaml;
+      return false;
+    }
+    Utils::writeFile(app_root / "docker-compose.yml", yaml);
+    return true;
+  }
+
+  bool start() {
+    auto bin = boost::filesystem::canonical(compose_bin).string();
+    std::string cmd("cd " + app_root.string() + " && " + bin + " up --remove-orphans -d");
+    if (std::system(cmd.c_str()) != 0) {
+      return false;
+    }
+    return true;
+  }
+
+  std::string name;
+  boost::filesystem::path app_root;
+  boost::filesystem::path app_params;
+  boost::filesystem::path app_bin;
+  boost::filesystem::path compose_bin;
+};
+
+bool DockerAppManager::iterate_apps(const Uptane::Target &target, DockerAppCb cb) const {
+  auto apps = target.custom_data()["docker_apps"];
+  bool res = true;
+  Uptane::ImagesRepository repo;
+  repo.checkMetaOffline(*storage_);
```

Comment:
`checkMetaOffline` pulls in data from `INvStorage` to properly initialize the `targets` member. It could be avoided by getting the SotaUptaneClient to pass its copy of that variable, but when I tried that approach I didn't really like the way to code looked. I'm open to trying it that way again and letting you decide what you prefer.

Reply:
Yeah, that would require changing all of the package managers to support it. For now it's probably fine; it just seems like an indirect way to get what you are looking for. Again, a comment to explain why the check is useful might be helpful.